plugins/formal-verification/src/org/jetbrains/kotlin/formver/embeddings/ExpEmbedding.kt

```diff
@@ -0,0 +1,187 @@
+
+sealed interface ExpEmbedding {
+    val type: TypeEmbedding
+    val viperExp: Exp
+
+    fun withType(newType: TypeEmbedding): ExpEmbedding =
+        when {
+            type == newType -> this
+            type !is NullableTypeEmbedding && newType is NullableTypeEmbedding -> NullableOf(this.withType(newType.elementType))
+            // This NullLit conversion is just hard-coded here to not break the tests,
+            // it can be deleted once the to-do below has been solved.
+            this is NullLit && newType is NullableTypeEmbedding -> NullLit(newType)
+            type is NullableTypeEmbedding && (type as NullableTypeEmbedding).elementType.isSubTypeOf(newType) ->
+                ValOfNullable(this).withType(newType)
+            type is NullableTypeEmbedding && newType is NullableTypeEmbedding -> TODO("Add domain function in the Nullable domain to convert from one type of nullable to another if their element types are subtypes.")
+            type.isSubTypeOf(newType) -> throw NotImplementedError("Type $type is a subtype of $newType but no conversion function was specified.")
+            else -> throw IllegalArgumentException("Expression $this of type $type is not a subtype of $newType and can thus not be converted to it.")
+        }
+}
+
+fun <E : ExpEmbedding> List<E>.viperExps(): List<Exp> = map { it.viperExp }
+
+data class Add(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.Add =
+        Exp.Add(left.withType(type).viperExp, right.withType(type).viperExp)
+}
+
+data class Sub(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.Sub =
+        Exp.Sub(left.withType(type).viperExp, right.withType(type).viperExp)
+}
+
+data class Mul(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.Mul =
+        Exp.Mul(left.withType(type).viperExp, right.withType(type).viperExp)
+}
+
+data class Div(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.Div =
+        Exp.Div(left.withType(type).viperExp, right.withType(type).viperExp)
+}
+
+data class Mod(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.Mod =
+        Exp.Mod(left.withType(type).viperExp, right.withType(type).viperExp)
+}
+
+data class LtCmp(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = BooleanTypeEmbedding
+    val cmpArgType = commonSuperType(left.type, right.type)
```

Comment:
Is it worth making an abstract class or something for this so that this pattern can be deduplicated?